WebRequestError when performing request without connectivity

Fenix: https://github.com/mozilla-mobile/fenix/issues/520

```
 E  Process: org.mozilla.fenix, PID: 8203
 E  org.mozilla.geckoview.WebRequestError: Request failed, error=37, category=5
 E      at org.mozilla.gecko.mozglue.GeckoLoader.nativeRun(Native Method)
 E      at org.mozilla.gecko.GeckoThread.run(GeckoThread.java:81)
```

Happens when the device has no connectivity (e.g. airplane mode) and we perform a request with `GeckoViewFetchClient`.

--- a/components/browser/engine-gecko-nightly/src/androidTest/java/mozilla/components/browser/engine/gecko/fetch/geckoview/GeckoViewFetchTestCases.kt
+++ b/components/browser/engine-gecko-nightly/src/androidTest/java/mozilla/components/browser/engine/gecko/fetch/geckoview/GeckoViewFetchTestCases.kt
@@ -127,4 +127,10 @@ class GeckoViewFetchTestCases : mozilla.components.tooling.fetch.tests.FetchTest
     override fun get200WithCacheControl() {
         super.get200WithCacheControl()
     }
+
+    @Test
+    @UiThreadTest
+    override fun getThrowsIOExceptionWhenHostNotReachable() {
+        super.getThrowsIOExceptionWhenHostNotReachable()
+    }
 }

--- a/components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/fetch/GeckoViewFetchClient.kt
+++ b/components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/fetch/GeckoViewFetchClient.kt
@@ -17,6 +17,7 @@ import org.mozilla.geckoview.GeckoWebExecutor
 import org.mozilla.geckoview.WebRequest
 import org.mozilla.geckoview.WebRequest.CACHE_MODE_DEFAULT
 import org.mozilla.geckoview.WebRequest.CACHE_MODE_RELOAD
+import org.mozilla.geckoview.WebRequestError
 import org.mozilla.geckoview.WebResponse
 import java.io.IOException
 import java.io.InputStream
@@ -62,6 +63,8 @@ class GeckoViewFetchClient(
             return webResponse?.toResponse() ?: throw IOException("Fetch failed with null response")
         } catch (e: TimeoutException) {
             throw SocketTimeoutException()
+        } catch (e: WebRequestError) {
+            throw IOException(e)
         }
     }
 

--- a/components/browser/engine-gecko-nightly/src/test/java/mozilla/components/browser/engine/gecko/fetch/GeckoViewFetchUnitTestCases.kt
+++ b/components/browser/engine-gecko-nightly/src/test/java/mozilla/components/browser/engine/gecko/fetch/GeckoViewFetchUnitTestCases.kt
@@ -29,6 +29,7 @@ import org.mozilla.geckoview.GeckoResult
 import org.mozilla.geckoview.GeckoRuntime
 import org.mozilla.geckoview.GeckoWebExecutor
 import org.mozilla.geckoview.WebRequest
+import org.mozilla.geckoview.WebRequestError
 import org.mozilla.geckoview.WebResponse
 import org.robolectric.RobolectricTestRunner
 import java.io.IOException
@@ -271,6 +272,7 @@ class GeckoViewFetchUnitTestCases : mozilla.components.tooling.fetch.tests.Fetch
         assertEquals("ÄäÖöÜü", response.body.string())
     }
 
+    @Test
     override fun get200WithCacheControl() {
         mockResponse(200)
 
@@ -284,6 +286,15 @@ class GeckoViewFetchUnitTestCases : mozilla.components.tooling.fetch.tests.Fetch
 
         verify(geckoWebExecutor)!!.fetch(captor.capture(), eq(GeckoWebExecutor.FETCH_FLAGS_NONE))
         assertEquals(WebRequest.CACHE_MODE_RELOAD, captor.value.cacheMode)
+    }
+
+    @Test(expected = IOException::class)
+    override fun getThrowsIOExceptionWhenHostNotReachable() {
+        val executor = mock(GeckoWebExecutor::class.java)
+        `when`(executor.fetch(any(), anyInt())).thenAnswer { throw WebRequestError(0, 0) }
+        geckoWebExecutor = executor
+
+        createNewClient().fetch(Request(""))
     }
 
     private fun mockRequest(headerMap: Map<String, String>? = null, body: String? = null, method: String = "GET") {

--- a/components/tooling/fetch-tests/src/main/java/mozilla/components/tooling/fetch/tests/FetchTestCases.kt
+++ b/components/tooling/fetch-tests/src/main/java/mozilla/components/tooling/fetch/tests/FetchTestCases.kt
@@ -472,6 +472,23 @@ abstract class FetchTestCases {
         assertEquals("Could've cached this!", response2.body.string())
     }
 
+    @Test
+    open fun getThrowsIOExceptionWhenHostNotReachable() {
+        try {
+            val client = createNewClient()
+            val response = client.fetch(Request(url = "http://invalid.offline"))
+
+            // We're doing this the old-fashioned way instead of using the
+            // expected= attribute, because the test is launched on a different
+            // thread (using a different coroutine context) than this block.
+            fail("Expected IOException, but got response: ${response.status}")
+        } catch (e: IOException) {
+            // expected
+        } catch (e: Exception) {
+            fail("Expected IOException")
+        }
+    }
+
     private inline fun withServerResponding(
         vararg responses: MockResponse,
         crossinline block: MockWebServer.(Client) -> Unit

